Adds decodable support

pogramos opened the pull request:
This PR removes `Unbox` in favor of using 'Decodable' on serializations.

It will also remove protocol extensions where Value conforms to `Collection` or `Self` to `JSONEnvelope`.

`Value` to `Collection` was causing lots of ambiguities since `Collection` is also a Decodable object

If there's a necessity to serialize more complex structures, the resource can always implement

```swift
func serialize(response: Result<Data>) -> SerializationResult<Value>
```

e.g.:
```swift
func seiralize(response: Result<Data>) -> SerializationResult<Value> {
  switch response {
  case let .success(data):
    do {
       // deserialize and return .success(<your object result>)
    } catch {
      return .failure(error)
    }
  case let .failure(error):
    return .failure(error)
  }
}
```

It also adds `Swift Package Manager` support for this version

Commit:
improvement(networking+decoding): adds new extension to use decodable instead of unbox

pogramos committed:
improvement(Decoding): proposes new decoding logic

This commit will remove `Unbox` in favor of using 'Decodable' on

serializations.

It will also remove protocol extensions where Value conforms to Collection

or `Self` to `JSONEnvelope`.

I don't think it is necessary since the object can use

```swift
func serialize(response:)
```

its own method to serialize more complex structures and conforming

Value to collection was causing lots of ambiguities, since `Collection` is also a Decodable object

BREAKING CHANGE: Removes Unbox, Updates decoding logic

pogramos committed:
improvement(build): WIP: Adds package.swift to the project

pogramos committed:
improvement(wip): changing testable modules

Commit:
test: updates decoding tests o use 8.0.0 and 9.0.0 version of OHHTTPStubs

This was necessary since only 9.0.0 version of OHHTTPStubs supports SPM

BREAKING CHANGE: Adds new version of OHHTTPStubs to the tests

Commit:
chore: bump version to 4.0.0

fellipecaetano commented on Source/Decoding/Networking+Decoding.swift:
```diff
@@ -0,0 +1,32 @@
+    }
+}
+
+private extension JSONSerialization {
```
I didn't understand the point of this extension.

fellipecaetano commented on Source/Info.plist:
```diff
@@ -15,7 +15,7 @@
 	<string>FMWK</string>
 	<key>CFBundleShortVersionString</key>
-	<string>3.0.0</string>
+	<string>$(MARKETING_VERSION)</string>
```
I think the `$(MARKETING_VERSION)` pattern is nice in theory but in practice it causes some obscure problems. For instance, there are Bitrise workflow steps that are able to extract Xcode project information into variables that can be used by the rest of the workflow. If we use `$(MARKETING_VERSION)` instead of the raw build number, Bitrise is not able to determine what's the actual build number and we get some funky behavior as a consequence.

fellipecaetano commented on Tests/Cache/NetworkingTests+Cache.swift:
```diff
@@ -201,6 +208,15 @@ class CacheNetworkingTests: XCTestCase {
     }
+
+
+    private func systemPath(_ resource: String, extension ext: String) -> URL? {
```
I think the `systemPath(_:extension:)` name is misleading since it might mean the path is for a resource in the user's machine, which is not necessarily true. Maybe `resourcePath(_:extension:)` would be better?

fellipecaetano commented on Tests/Decoding/NetworkingTests+Decoding.swift:
```diff
@@ -1,19 +1,28 @@
     func testSuccessfulRequestForObject() {
         _ = stub(condition: isPath("/song")) { _ in
-            return fixture(filePath: OHPathForFile("song.json", type(of: self))!, headers: nil)
+
```
I think we can do away with this blank line.

fellipecaetano commented on Tests/Fakes/Song.swift:
```diff
@@ -70,11 +64,28 @@ extension Song {
             }
         }
+
+        func serialize(response: Result<Data>) -> SerializationResult<[Song]> {
```
One of Wolf's design objectives was to eliminate (most of) the need of manually implementing this `serialize(response:)` method. Can we achieve the same behavior with `Decodable`?

pogramos replied:
I had some problems trying to match the extension with the `Decodable` alongside with `Collection`, but i've refactored the serialize code to achieve the expected result, check it out and tell me what you think

fellipecaetano commented on Tests/Fixtures/OHHTPStubs+Extension.swift:
```diff
@@ -0,0 +1,39 @@
+import Foundation
```
I think the name of this file could be more descriptive. Something like `+Paths` or `+Files` or maybe even `+Filesystem`.

Commit:
refactor(networking+decoding.swift): removing unused extension

Commit:
refactor: applying pr suggestions

Commit:
refactor: applying pr suggestions

Commit:
improvement(decoding): accepts complex structures with root keys

removes unused fixtures
refactoring the serialize to accept rootedKey objects

Commit:
chore(ohhtpstubs+filesystem.swift): renaming extension

pogramos committed:
build(package): updating package.swift to the latest version of the CwlPreconditionTesting

pogramos committed:
Merge branch 'master' of github.com:pogramos/Wolf

pogramos committed:
chore(package): updating nimble to master

version 8.0.5 had some problems loading CwlPreconditionTesting so, pointing it to master solves the
problem

pogramos commented:
I recommend waiting for `Nimble` to close this [685](https://github.com/Quick/Nimble/issues/685) milestone to merge this PR since there's a lot going on with their dependency on CwlPreconditionTesting framework for swift 5+ and It would be good to reference a tagged version instead of `.branch("master")` on SPM

pogramos committed:
fix(dependencies): updates dependencies

Fixing messy dependency bug on CwlPreconditionTesting from Nimble when using it as a Swift Package

pogramos committed:
chore(carthage): updating carthage dependencies

pogramos commented:
Issues solved with the new release of Nimble 
https://github.com/Quick/Nimble/pull/729

Commit:
improvement(dependencies): updating dependencies

Commit:
improvement(dependencies): updating podfile specs

Commit:
docs(podspec): updating podspec version

updating a version and deployment target on podspec
